x-pack/plugins/triggers_actions_ui/public/application/sections/alerts_table/hooks/use_columns.ts

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { EuiDataGridColumn } from '@elastic/eui';
+import { IStorageWrapper } from '@kbn/kibana-utils-plugin/public';
+import { AlertConsumers } from '@kbn/rule-data-utils';
+import { BrowserField, BrowserFields } from '@kbn/rule-registry-plugin/common';
+import { useCallback, useEffect, useState } from 'react';
+import { AlertsTableStorage } from '../alerts_table_state';
+import { useFetchBrowserFieldCapabilities } from './use_fetch_browser_fields_capabilities';
+
+interface UseColumnsArgs {
+  featureIds: AlertConsumers[];
+  storageAlertsTable: React.MutableRefObject<AlertsTableStorage>;
+  storage: React.MutableRefObject<IStorageWrapper>;
+  id: string;
+  defaultColumns: EuiDataGridColumn[];
+}
+
+const fieldTypeToDataGridColumnTypeMapper = (fieldType: string | undefined) => {
+  if (fieldType === 'date') return 'datetime';
+  if (fieldType === 'number') return 'numeric';
+  if (fieldType === 'object') return 'json';
+  return fieldType;
+};
+
+/**
+ * EUI Data Grid expects the columns to have a property 'schema' defined for proper sorting
+ * this schema as its own types as can be check out in the docs so we add it here manually
+ * https://eui.elastic.co/#/tabular-content/data-grid-schema-columns
+ */
+const euiColumnFactory = (
+  column: EuiDataGridColumn,
+  browserFields: BrowserFields
+): EuiDataGridColumn => {
+  const browserFieldsProps = getBrowserFieldProps(column.id, browserFields);
+  return {
+    ...column,
+    schema: fieldTypeToDataGridColumnTypeMapper(browserFieldsProps.type),
+  };
+};
+
+/**
+ * Searches in browser fields object for a specific field
+ */
+const getBrowserFieldProps = (
+  _columnId: string,
+  browserFields: BrowserFields
+): Partial<BrowserField> => {
+  for (const [, categoryDescriptor] of Object.entries(browserFields)) {
+    if (!categoryDescriptor.fields) {
+      continue;
+    }
+
+    for (const [fieldName, fieldDescriptor] of Object.entries(categoryDescriptor.fields)) {
+      if (fieldName === _columnId) {
+        return fieldDescriptor;
+      }
+    }
+  }
+  return { type: 'string' };
+};
+
+/**
+ * @param columns Columns to be considered in the alerts table
+ * @param browserFields constant object with all field capabilities
+ * @returns columns but with the info needed by the data grid to work as expected, e.g sorting
+ */
+const populateColumns = (
+  columns: EuiDataGridColumn[],
+  browserFields: BrowserFields
+): EuiDataGridColumn[] => {
+  return columns.map((column: EuiDataGridColumn) => {
```

Comment:
Here we are doing an O(n*m) complexity search, for all columns and all fields (when calling `getBrowserFieldProps`). It could be improved to O(n), being n the number of fields in browserFields. 

I guess `columns` will stay in a relatively low number of items, but I think we should be foresighted on this, we have had performance problems here in the past